app/controllers/repos_controller.rb

```diff
@@ -98,53 +96,69 @@ def benchmark_runs_for(benchmark_type)
       .sort_by { |run| run.initiator.created_at }
   end
 
-  def comparing_runs_for(benchmark_type)
-    BenchmarkRun
-      .fetch_commit_benchmark_runs(@comparing_benchmark.category, benchmark_type, @display_count)
-      .sort_by { |run| run.initiator.created_at }
-  end
-
-  def build_chart(benchmark_runs, benchmark_type)
-    comparing_runs = comparing_runs_for(benchmark_type) if @comparing_benchmark.present?
-
-    chart_builder = ChartBuilder.new(
-      benchmark_runs,
-      benchmark_type,
-      comparing_runs
-    )
+  def build_chart(benchmark_type)
+    if @comparing_benchmark.present?
+      ComparisonChartBuilder.new(benchmark_type, [@benchmark, @comparing_benchmark])
+    else
+      benchmark_runs = benchmark_runs_for(benchmark_type)
+
+      chart_builder = ChartBuilder.new(
+        benchmark_runs,
+        benchmark_type,
+      )
+
+      chart_builder.build_columns do |benchmark_run|
+        environment = YAML.load(benchmark_run.environment)
+        commit = benchmark_run.initiator
+
+        version = {
+          commit_sha: commit.sha1[0..6],
+          commit_date: commit.created_at.to_s,
+          commit_message: commit.message.truncate(30)
+        }
+        # If there is more information about the environment, we add it to `version`
+        if environment.is_a?(Hash)
+          version.merge!(environment)
+        else
+          version[:environment] = environment
+        end
 
-    chart_builder.build_columns do |benchmark_run|
-      environment = YAML.load(benchmark_run.environment)
-      commit = benchmark_run.initiator
-
-      version = {
-        commit_sha: commit.sha1[0..6],
-        commit_date: commit.created_at.to_s,
-        commit_message: commit.message.truncate(30)
-      }
-      # If there is more information about the environment, we add it to `version`
-      if environment.is_a?(Hash)
-        version.merge!(environment)
-      else
-        version[:environment] = environment
+        version
       end
 
-      version
+      chart_builder
     end
-
-    chart_builder
   end
 
   def cache(chart, benchmark_type)
+    if @comparing_benchmark.present?
+      cache_comparison(chart, benchmark_type)
+    else
+      $redis.set(
+        @cache_keys[benchmark_type],
+        {
+          datasets: chart.columns,
+          versions: chart.categories
+        }.to_msgpack
+      )
+    end
+  end
+
+  def cache_comparison(chart, benchmark_type)
     $redis.set(
       @cache_keys[benchmark_type],
```

Comment:
It will override cache only for that particular comparison chart. I am setting up cache keys in before_filter.